src/HexagonOptimize.cpp

```diff
@@ -1079,11 +1081,64 @@ class OptimizePatterns : public IRMutator {
             // that they generate.
             internal_assert(op->args.size() == 3);
             return mutate(lower_lerp(op->args[0], op->args[1], op->args[2]));
+        } else if (op->is_intrinsic(Call::div_round_to_zero) &&
+                   !op->type.is_float() && op->type.is_vector()) {
+            internal_assert(op->args.size() == 2);
+            Expr a = op->args[0];
+            Expr b = op->args[1];
+            // Run bounds analysis to estimate the range of result.
+            Expr extent_upper = find_constant_bound(a / b, Direction::Upper, bounds);
+            const uint64_t *upper_bound = as_const_uint(extent_upper);
+
+            a = mutate(a);
+            b = mutate(b);
+            debug(1) << "Using long div: (num: " << a << "); (den: " << b << ")\n";
+            if (a.type().is_uint()) {
```

Comment:
How about refactoring this so it just always casts a and b, does an unsigned long division, and then does a sign fixup if needed? That way, signed and unsigned are sharing the same division codepath.

Reply:
Also, I think this logic should be added in a `long_div` function, which calls `unsigned_long_div(abs(a), abs(b), ...)` and does the sign fixup if needed (for signed arguments).

Reply:
Done